### DISCUSSION TITLE
The length of ratio_list should be the same as the file_list.

### DISCUSSION THREAD

**Discussion by bound123** [PaddlePaddle/PaddleOCR#14389] [Q&A]
### 🔎 Search before asking

- [X] I have searched the PaddleOCR [Docs](https://paddlepaddle.github.io/PaddleOCR/) and found no similar bug report.
- [X] I have searched the PaddleOCR [Issues](https://github.com/PaddlePaddle/PaddleOCR/issues) and found no similar bug report.
- [X] I have searched the PaddleOCR [Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions) and found no similar bug report.

### 🐛 Bug (问题描述)

Traceback (most recent call last):
  File "D:\PaddleOCR-main\tools\train.py", line 269, in <module>
    main(config, device, logger, vdl_writer, seed)
  File "D:\PaddleOCR-main\tools\train.py", line 55, in main
    train_dataloader = build_dataloader(config, "Train", device, logger, seed)
  File "D:\PaddleOCR-main\ppocr\data\__init__.py", line 107, in build_dataloader
    dataset = eval(module_name)(config, mode, logger, seed)
  File "D:\PaddleOCR-main\ppocr\data\simple_dataset.py", line 42, in __init__
    assert (
AssertionError: The length of ratio_list should be the same as the file_list.

### 🏃‍♂️ Environment (运行环境)

python 3.10

### 🌰 Minimal Reproducible Example (最小可复现问题的Demo)

配置文件如图
Global:
  use_gpu: false
  epoch_num: 1200
  log_smooth_window: 20
  print_batch_step: 2
  save_model_dir: ./output/ch_db_res18/
  save_epoch_step: 1200
  # evaluation is run every 5000 iterations after the 4000th iteration
  eval_batch_step: [3000, 2000]
  cal_metric_during_train: False
  pretrained_model: D:\ancoda\anocada3\envs\paddle\Lib\site-packages\Preliminary_training\ch_ppocr_server_v2.0_det_train\best_accuracy.pdparams
  checkpoints:
  save_inference_dir:
  use_visualdl: False
  infer_img: doc/imgs_en/img_10.jpg
  save_res_path: ./output/det_db/predicts_db.txt

Architecture:
  model_type: detc
  algorithm: DB
  Transform:
  Backbone:
    name: ResNet_vd
    layers: 18
    disable_se: True
  Neck:
    name: DBFPN
    out_channels: 256
  Head:
    name: DBHead
    k: 50

Loss:
  name: DBLoss
  balance_loss: true
  main_loss_type: DiceLoss
  alpha: 5
  beta: 10
  ohem_ratio: 3

Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Cosine
    learning_rate: 0.001
    warmup_epoch: 2
  regularizer:
    name: 'L2'
    factor: 0

PostProcess:
  name: DBPostProcess
  thresh: 0.3
  box_thresh: 0.6
  max_candidates: 1000
  unclip_ratio: 1.5

Metric:
  name: DetMetric
  main_indicator: hmean

Train:
  dataset:
    name: SimpleDataSet
    data_dir: D:\ancoda\anocada3\envs\paddle\Lib\site-packages\train_data\det\train
    label_file_list:
     D:/ancoda/anocada3/envs/paddle/Lib/site-packages/train_data/det/train.txt
    ratio_list: [1.0]
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - DetLabelEncode: # Class handling label
      - IaaAugment:
          augmenter_args:
            - { 'type': Fliplr, 'args': { 'p': 0.5 } }
            - { 'type': Affine, 'args': { 'rotate': [-10, 10] } }
            - { 'type': Resize, 'args': { 'size': [0.5, 3] } }
      - EastRandomCropData:
          size: [960, 960]
          max_tries: 50
          keep_ratio: true
      - MakeBorderMap:
          shrink_ratio: 0.4
          thresh_min: 0.3
          thresh_max: 0.7
      - MakeShrinkMap:
          shrink_ratio: 0.4
          min_text_size: 8
      - NormalizeImage:
          scale: 1./255.
          mean: [0.485, 0.456, 0.406]
          std: [0.229, 0.224, 0.225]
          order: 'hwc'
      - ToCHWImage:
      - KeepKeys:
          keep_keys: ['image', 'threshold_map', 'threshold_mask', 'shrink_map', 'shrink_mask'] # the order of the dataloader list
  loader:
    shuffle: True
    drop_last: False
    batch_size_per_card: 8
    num_workers: 4

Eval:
  dataset:
    name: SimpleDataSet
    data_dir: D:\ancoda\anocada3\envs\paddle\Lib\site-packages\train_data\det\val
    label_file_list:
     D:/ancoda/anocada3/envs/paddle/Lib/site-packages/train_data/det/val.txt
    transforms:
      - DecodeImage: # load image
          img_mode: BGR
          channel_first: False
      - DetLabelEncode: # Class handling label
      - DetResizeForTest:
#           image_shape: [736, 1280]
      - NormalizeImage:
          scale: 1./255.
          mean: [0.485, 0.456, 0.406]
          std: [0.229, 0.224, 0.225]
          order: 'hwc'
      - ToCHWImage:
      - KeepKeys:
          keep_keys: ['image', 'shape', 'polys', 'ignore_tags']
  loader:
    shuffle: False
    drop_last: False
    batch_size_per_card: 1 # must be 1
    num_workers: 2

**Comment by GreatV** [comment]
The error message `AssertionError: The length of ratio_list should be the same as the file_list` indicates that there is a mismatch between the number of elements in `ratio_list` and `label_file_list` in your configuration file.

### Explanation
The `ratio_list` is used to specify the proportion of data to be used from each dataset listed in `label_file_list`. For every dataset in `label_file_list`, there must be a corresponding entry in `ratio_list` that defines its ratio. If the lengths of these two lists do not match, this error is raised.

In your configuration:
```yaml
label_file_list:
  D:/ancoda/anocada3/envs/paddle/Lib/site-packages/train_data/det/train.txt
ratio_list: [1.0]
```
Here, the `label_file_list` contains **one file**, and the `ratio_list` correctly contains **one ratio**, so this part seems fine. However, this error can also occur if:
1. **Your YAML formatting is incorrect.** YAML is sensitive to formatting, so make sure there are no indentation or parsing issues.
2. **The `label_file_list` is being parsed incorrectly.** The file path may not be recognized as a valid list, or the configuration parser may be misinterpreting it.

### Steps to Fix the Error

1. **Ensure Proper YAML Syntax**
   - Double-check the indentation and syntax in your configuration file. For example, ensure `label_file_list` is written as a proper list:
     ```yaml
     label_file_list:
       - D:/ancoda/anocada3/envs/paddle/Lib/site-packages/train_data/det/train.txt
     ratio_list: [1.0]
     ```
     Note the use of `-` before the file path to explicitly define it as a list.

2. **Verify the Dataset File**
   - Ensure the file `D:/ancoda/anocada3/envs/paddle/Lib/site-packages/train_data/det/train.txt` exists and is accessible. If the file is missing, the `label_file_list` might be treated as empty.

3. **Check for Multiple Datasets**
   - If you plan to use multiple datasets in `label_file_list`, ensure the `ratio_list` has the same number of elements. For example:
     ```yaml
     label_file_list:
       - path/to/dataset1.txt
       - path/to/dataset2.txt
     ratio_list: [0.6, 0.4]
     ```
     Here, `ratio_list` specifies that 60% of the data should come from `dataset1` and 40% from `dataset2`.

4. **Debug the Configuration Parsing**
   - Insert debug statements or print the parsed configuration to verify what `label_file_list` and `ratio_list` contain at runtime. For example, you can add a print statement in the script (`simple_dataset.py`) where the assertion is failing:
     ```python
     print(f"label_file_list: {self.label_file_list}")
     print(f"ratio_list: {self.ratio_list}")
     ```

5. **Update PaddleOCR Version**
   - Ensure you are using the latest stable version of PaddleOCR and PaddlePaddle. Bugs in earlier versions might cause unexpected parsing issues. To update:
     ```bash
     pip install --upgrade paddleocr paddlepaddle
     ```

6. **Cross-Check with Related Discussions**
   - The linked [Discussion](https://github.com/PaddlePaddle/PaddleOCR/discussions/13811) explains issues related to dataset configuration and shape mismatches, which might also cause similar errors. Ensure your dataset is properly pre-processed and the file paths are correct.

### Additional Notes
- If you are still facing issues after following the steps above, try running a minimal configuration with a dummy dataset to isolate the problem. Test with:
  ```yaml
  label_file_list:
    - dummy_dataset.txt
  ratio_list: [1.0]
  ```
  Ensure `dummy_dataset.txt` contains valid data in the expected format.

By properly aligning the lengths of `label_file_list` and `ratio_list`, and ensuring the dataset files are correctly specified, you should be able to resolve this error.

---
*Response generated by [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*
